#+TITLE: Cool Bioinformatics Scripts

* [[file:qqplot.py][qqplot]]
You can use this script in two ways

- read tons of millions p values from stdin

#+begin_src shell
zcat pval.txt.gz | qqplot.py -out test -title "QQ plot on the fly"
#+end_src

- use qqplot.py in your script

#+begin_src python
import numpy as np
from qqplot import qqplot
p = np.random.random(1000000)
qqplot(x=p, figname="test.png")
#+end_src

* [[file:fixref.py][fixref]]

Before running =bcftools merge=, you maybe need to fix the ref and alt and corresponding genotypes, otherwise =bcftools= will surprise you.

#+begin_src shell
usage: fixref.py [-h] REF_VCF IN_VCF OUT_VCF
#+end_src

* [[file:calc_imputed_gt_discord.py][calculate genotype discordance]]

When you run imputation analysis with =BEAGLE= (or other imputation tools), you may want to know the distribution of genotype discordance between the original vcf and imputed vcf.

*warning* : Before running the script, you must be sure the two vcfs have the exact same sites and samples for each chromosome.

#+begin_src shell
usage: calc_imputed_gt_discord.py [-h] [-chr STRING] VCF1 VCF2 OUT
#+end_src

[[file:image/calc_imputed_gt_discord.png][a output png]]
